flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -144,12 +145,258 @@ mlir::Value calculateTripCount(fir::FirOpBuilder &builder, mlir::Location loc,
 } // namespace Fortran
 
 namespace {
+namespace looputils {
+struct InductionVariableInfo {
+  mlir::Operation *iterVarMemDef;
+  llvm::SetVector<mlir::Operation *> indVarUpdateOps;
+};
+
+using LoopNestToIndVarMap =
+    llvm::MapVector<fir::DoLoopOp, InductionVariableInfo>;
+
+/// Collect the list of values used inside the loop but defined outside of it.
+/// The first item in the returned list is always the loop's induction
+/// variable.
+void collectLoopLiveIns(fir::DoLoopOp doLoop,
+                        llvm::SmallVectorImpl<mlir::Value> &liveIns) {
+  // Given an operation `op`, this lambda returns true if `op`'s operand is
+  // ultimately the loop's induction variable. Detecting this helps finding
+  // the live-in value corresponding to the induction variable in case the
+  // induction variable is indirectly used in the loop (e.g. throught a cast
+  // op).
+  std::function<bool(mlir::Operation * op)> isIndVarUltimateOperand =
```

Comment:
This doesn't need to be a lambda function. For reuse purposes it is better to make it a static function.